Research/QuantBook.cs

```diff
@@ -144,62 +175,122 @@ public QuantBook() : base()
         }
 
         /// <summary>
-        /// Get fundamental data from given symbols
+        /// Python implementation of GetFundamental, get fundamental data for input symbols or tickers
         /// </summary>
-        /// <param name="pyObject">The symbols to retrieve fundamental data for</param>
+        /// <param name="input">The symbols or tickers to retrieve fundamental data for</param>
         /// <param name="selector">Selects a value from the Fundamental data to filter the request output</param>
         /// <param name="start">The start date of selected data</param>
         /// <param name="end">The end date of selected data</param>
-        /// <returns></returns>
-        public PyObject GetFundamental(PyObject tickers, string selector, DateTime? start = null, DateTime? end = null)
+        /// <returns>pandas DataFrame</returns>
+        public PyObject GetFundamental(PyObject input, string selector, DateTime? start = null, DateTime? end = null)
         {
+            //Null selector is not allowed for Python DataFrame
             if (string.IsNullOrWhiteSpace(selector))
             {
-                return "Invalid selector. Cannot be None, empty or consist only of white-space characters".ToPython();
+                throw new ArgumentException("Invalid selector. Cannot be None, empty or consist only of white-space characters");
             }
 
+            //Covert to symbols
+            var symbols = PythonUtil.ConvertToSymbols(input);
+            
+            //Fetch the data
+            var fundamentalData = GetAllFundamental(symbols, selector, start, end);
+
             using (Py.GIL())
             {
-                // If tickers are not a PyList, we create one
-                if (!PyList.IsListType(tickers))
+                var data = new PyDict();
+                foreach (var day in fundamentalData.OrderBy(x => x.Key))
                 {
-                    var tmp = new PyList();
-                    tmp.Append(tickers);
-                    tickers = tmp;
+                    var orderedValues = day.Value.OrderBy(x => x.Key.ID.ToString());
+                    var columns = orderedValues.Select(x => x.Key.ID.ToString());
+                    var values = orderedValues.Select(x => x.Value);
```

Comment:
We should double-check these two selects are not causing the OrderBy to happen twice, maybe it has a cache? 🤔

Reply:
True. I think it might very well be happening twice so I will use orderedValues as a list instead to avoid the possible multiple enumeration.